lint/rule_panel_rate_interval.go

```diff
@@ -2,50 +2,70 @@ package lint
 
 import (
 	"fmt"
-	"regexp"
+	"time"
+
+	"github.com/prometheus/prometheus/promql/parser"
 )
 
-var rangeVectorRegexp = regexp.MustCompile(`\[([^:)]+)\]`)
+type inspector func(parser.Node, []parser.Node) error
+
+func (f inspector) Visit(node parser.Node, path []parser.Node) (parser.Visitor, error) {
+	if err := f(node, path); err != nil {
+		return nil, err
+	}
+	return f, nil
+}
 
 // NewPanelRateIntervalRule builds a lint rule for panels with Prometheus queries which checks
 // all range vector selectors use $__rate_interval.
 func NewPanelRateIntervalRule() *PanelRuleFunc {
+	rateIntervalMagicDuration, err := time.ParseDuration(globalVariables["__rate_interval"].(string))
+	if err != nil {
+		// Will not happen
+		panic(err)
+	}
 	return &PanelRuleFunc{
 		name:        "panel-rate-interval-rule",
 		description: "Checks that each panel uses $__rate_interval.",
 		fn: func(d Dashboard, p Panel) Result {
 			if t := getTemplateDatasource(d); t == nil || t.Query != "prometheus" {
 				// Missing template datasources is a separate rule.
-				return Result{
-					Severity: Success,
-					Message:  "OK",
-				}
+				return SuccessResult
 			}
 
 			if !panelHasQueries(p) {
 				// Don't lint certain types of panels.
-				return Result{
-					Severity: Success,
-					Message:  "OK",
-				}
+				return SuccessResult
 			}
 
 			for _, target := range p.Targets {
-				// Hack in replace [$__rate_interval] with [5m] so queries parse correctly.
-				for _, rate := range rangeVectorRegexp.FindAllString(target.Expr, -1) {
-					if rate != "[$__rate_interval]" {
-						return Result{
-							Severity: Error,
-							Message:  fmt.Sprintf("Dashboard '%s', panel '%s' invalid PromQL query '%s': should use $__rate_interval", d.Title, p.Title, target.Expr),
-						}
+				expr, err := parsePromQL(target.Expr, d.Templating.List)
+				if err != nil {
+					continue
+				}
+				err = parser.Walk(inspector(func(node parser.Node, parents []parser.Node) error {
+					selector, ok := node.(*parser.MatrixSelector)
+					if !ok || selector.Range == rateIntervalMagicDuration || len(parents) == 0 {
+						// We are not inspecting something like foo{...}[...] or the range vector selector is $__rate_interval
+						return nil
+					}
```

Comment:
I think I'd prefer slightly more verbose if statements; putting 3 different clauses in gets confusing for me :-)

```suggestion
					selector, ok := node.(*parser.MatrixSelector)
					if !ok {
						// We are not inspecting something like foo{...}[...]
						continue
					}
					
					 
					if selector.Range == rateIntervalMagicDuration {
						 // Range vector selector is $__rate_interval
						 return nil
					}
					 
					 if len(parents) == 0 {
						// Bit weird to have a naked foo[$__rate_interval], but allow it.
						return nil
					}
```